content: draft: further refinement for revision continuity (#1381)
slsa-framework/slsa: docs/spec/draft/source-requirements.md

`refs/heads/releases/*`, but not `refs/heads/playground/*`.

<td><td>✓<td>✓<td>✓
<tr id="continuity"><td>Continuity<td>

Revisions are created by applying a specific code change (a "diff" in git) on

> top of an earlier revision of a branch. It MUST NOT be possible to rewrite the
> history of revisions for protected branches.

Reviewer: ```suggestion
top of an earlier revision of a branch. The SCS MUST prevent tampering with the history of revisions on protected branches.
```

In other words, if the organization updates a branch from commit A to commit B,
commit B MUST be a descendant of A. For systems like GitHub or GitLab, this can
be accomplished by enabling branch protection rules that prevent force pushes
and branch deletions.

It MUST NOT be possible to delete the entire repository (including all branches)

> and replace it with different source.

Author: The rebase had conflicts but I took this content when performing it.

> Continuity MUST be established and tracked from a specific revision. If there is

Reviewer 2: We might need to keep the branch qualifier when talking about the revision.  A revision can be created elsewhere on a non-protected branch and then brought into a protected branch.

Perhaps

```suggestion
Continuity MUST be established and tracked from a specific revision within the context of the protected branch. If there is
```

or (if we're going to accept time)

```suggestion
Continuity MUST be established and tracked from a specific revision (or the time associated with that revision) within the context of the protected branch. If there is
```

Reviewer: TLDR, I think the wording is ok as-is. 

source provenance _can_ be written as soon as the revision is uploaded. 
in practice, that's just too much, so it will often only be emitted for "special" revisions that become pointed-to by branches.

It should be ok for a revision to "lose" reachability from a branch and successfully reestablish it without breaking everything. "oops I deleted the main branch" is the critical scenario there.

Reviewer 2: Zac your suggestion we just resolve this thread and not take the suggestions?

My concern was that many controls relate to the branch itself, but I suppose they don't _all_ need to and that's perhaps control specific.

If so, then yeah maybe let's just resolve this thread and see how it goes.

> Continuity MUST be established and tracked from a specific revision. If there is

Author: Would we (for example, in the POC) be able to make an assumption about controls being in place since _a revision_ if we have checks to ensure that the controls are in place since _that revisions' time_?

That is, can a SCS optionally use its time of a revision and the policies in place as a proxy for ensuring that controls have been in place from a given revision?

> Continuity MUST be established and tracked from a specific revision. If there is

Reviewer: >  time of a revision

It seemed that there were a bunch of reasons to remove time from the spec itself. 
I think for the POC based on current GH tooling, you should be ok saying "the apis are _rough_ so I'm using timestamp to identify the revision needed by the spec."

> Continuity MUST be established and tracked from a specific revision. If there is

Reviewer 2: I'm convinced.

The PoC can say something like

> "Continuity MUST be established and tracked from a specific revision" is met by tracking continuity from the time the revision was introduced to a given context/branch.

If that sounds right to everyone we can drop the time related comments.

> Continuity MUST be established and tracked from a specific revision. If there is

Author: I think that sounds reasonable for the PoC.

a lapse in continuity for a specific control, that continuity MUST be
re-established from a new revision.

> Continuity MUST also apply to any [change management processes](#enforced-change-management-process)

Reviewer: Is there a difference between change management controls and other controls? 
Seems like this section is well covered by the above 

> a lapse in continuity for a specific control, that continuity MUST be
re-established from a new revision.

> Continuity MUST also apply to any [change management processes](#enforced-change-management-process)

Reviewer 2: oh that's a good point, it would be better to keep this simple.

Optional: If we wanted to make it more clear without the extra language it could be something like

> a lapse in continuity for a specific control (including those enforced by the change management process),

> Continuity MUST also apply to any [change management processes](#enforced-change-management-process)

Reviewer 2: Sounds like you'd like to push without, fine with me either way.

  phishing resistant.
-   Protect against authentication token theft by forbidding bearer tokens
  (e.g. PATs).

> -   Including length of continuity in the VSAs

Reviewer 2: If continuity is 'just' revision based the length might be difficult to include.  Perhaps the initial revision could be included instead, but that might be less meaningful to users?

Author: I wasn't trying to write out time from continuity as I mentioned above. I agree that it doesn't seem to make a lot of sense to have "continuity since revision" in the VSA, but that "continuity since time" is much more natural. Should we change this if we make updates above to indicate that time _may_ be a proxy for revision above?
